old-materials/model_viz.py

```diff
@@ -0,0 +1,608 @@
+import inspect
+import pydot
+import matplotlib.pyplot as plt
+import colorsys
+import random
+import os
+from collections import defaultdict
+from skagent.model import Control, DBlock, RBlock
+
+class DBlockVisualizer:
+    """
+    A tool class for visualizing DBlock model structure.
+    Extracts information from DBlock objects and generates logical flowcharts.
+    Supports identification of time dependencies to create a directed acyclic graph.
+    """
+    
+    def __init__(self, block, agent_attribution=None, calibration=None):
+        """
+        Initialize visualizer
+        
+        Parameters:
+        block -- DBlock object to visualize
+        agent_attribution -- Dictionary of variables belonging to different entities, e.g. {"consumer": [...], "lender": [...]}
+        calibration -- Model calibration parameters, used if provided
+        """
+        self.block = block
+        self.agent_attribution = agent_attribution or {}
+        self.calibration = calibration or {}
+        
+        # Store extracted information
+        self.variables = {}
+        self.dependencies = {}
+        self.formulas = {}
+        self.prev_period_vars = set()
+        
+        # Use calibration parameters as model parameters
+        self.parameters = self.calibration
+    
+    def analyze(self):
+        """
+        Analyze DBlock and extract all necessary information
+        """
+        self._identify_variables()
+        self._extract_dependencies()
+        self._identify_time_dependencies()
+        self._extract_formulas()
+        return self
+    
+    def _identify_variables(self):
+        """Identify and classify all variables in DBlock"""
+        variables = {
+            'shock_vars': [],     # Random shocks
+            'state_vars': [],     # State variables
+            'control_vars': [],   # Control variables
+            'reward_vars': [],    # Reward variables
+            'param_vars': set()   # Model parameters
+        }
+        
+        # Track all classified model variables to avoid duplicates
+        classified_vars = set()
+        
+        # 1) Extract random shock variables (if present)
+        if hasattr(self.block, 'shocks'):
```

Comment:
The blocks will always have a shocks attribute because it is filled with an empty dictionary by default.
Likewise for the other model attributes you're testing for with `hasattr` here.

https://github.com/scikit-agent/scikit-agent/blob/main/src/skagent/model.py#L231

These `hasattr` conditionals are not necessary, and they are also not very performant. It would be better to remove them.